# --- T2-COPYRIGHT-BEGIN ---
# t2/package/*/llvm/hotfix-ppc-jit.patch
# Copyright (C) 2025 The T2 SDE Project
# SPDX-License-Identifier: GPL-2.0 or patched project license
# --- T2-COPYRIGHT-END ---

root@t2:~# glxgears
LLVM ERROR: Relocation type 0x%x not implemented yet!
Aborted                    (core dumped) glxgears

Signed-off-by: Rene Rebe <rene@t2linux.com>

--- llvm-20.1.8.src/lib/ExecutionEngine/RuntimeDyld/RuntimeDyldELF.cpp.orig	2025-07-09 01:06:32.000000000 +0200
+++ llvm-20.1.8.src/lib/ExecutionEngine/RuntimeDyld/RuntimeDyldELF.cpp	2025-07-28 17:17:23.997152821 +0200
@@ -1028,7 +1030,13 @@
   case ELF::R_PPC_ADDR16_HA:
     writeInt16BE(LocalAddress, applyPPCha(Value + Addend));
     break;
-  }
+  case ELF::R_PPC_REL32: {
+    uint32_t FinalAddress = Section.getLoadAddressWithOffset(Offset);
+    int32_t delta = static_cast<int32_t>(Value - FinalAddress + Addend);
+    writeInt32BE(LocalAddress, delta);
+  } break;
+ }
+
 }
 
 void RuntimeDyldELF::resolvePPC64Relocation(const SectionEntry &Section,
